feat: add flexible routing system for invitation notifications

andrii-trush opened the pull request:
- Use Filament default routing when config is empty
- Support custom route configuration via filament-invite.routes.reset
- Maintain backward compatibility with existing setups

andrii-trush committed:
feat: add flexible routing system for invitation notifications

- Use Filament default routing when config is empty
- Support custom route configuration via filament-invite.routes.reset
- Maintain backward compatibility with existing setups

andrii-trush commented:
Hello @swilla @scottgrayson . Could You also review this PR? It should be useful especially when  `app.url` is not equal filament url

andrii-trush commented:
Also as my previous PR. I've tested only with Filament 3. Be careful with release please.

andrii-trush committed:
Merge branch 'TappNetwork:main' into filament-url-refactoring-for-notification

andreia commented on README.md:
````diff
@@ -68,6 +68,34 @@ Invite action outside of a table uses a different class
 
 ```
 
+## Notification Routing
+
+The package uses a simple routing strategy for password reset URLs in invitation emails:
+
+- **Default**: If `filament-invite.routes.reset` config is empty, uses Filament's default routing
+- **Custom**: If a route name is configured, uses the specified custom route
+
+```php
+if (empty(config('filament-invite.routes.reset'))) {
+    $url = Filament::getResetPasswordUrl($this->token, $email, ['invite' => true]);
+} else {
+    $domain = rtrim(Filament::getPanel($this->filamentPanelId)->getPath(), '/');
+    $url = url($domain . '/' . route(config('filament-invite.routes.reset'), [
+        'token' => $this->token,
+        'email' => $email,
+        'invite' => true,
+    ], false));
+}
+```
+
+To use custom routing, set your route name in the config:
+
+```php
+'routes' => [
+    'reset' => 'your.custom.route.name',
+],
+```
````
Updated README:

## Notification Routing

The password reset URL  can be customized or use the Filament's authentication system:

- **Default**: Uses Filament's built-in routing (respects panel's `passwordResetRouteSlug` configuration)
- **Custom**: Uses completely custom routes outside Filament's authentication system

```php
// Get the appropriate panel
$panel = $this->filamentPanelId
    ? Filament::getPanel($this->filamentPanelId)
    : Filament::getDefaultPanel();

if (empty(config('filament-invite.routes.reset'))) {
    // Use Filament's built-in routing (respects panel's passwordResetRouteSlug configuration)
    $url = $panel->getResetPasswordUrl($this->token, $notifiable, ['invite' => true]);
} else {
    // Support both custom routes and full URLs for maximum flexibility
    $customRoute = config('filament-invite.routes.reset');

    // Check if it's a full URL (starts with http/https)
    if (str_starts_with($customRoute, 'http')) {
        // Use the full URL directly
        $url = $customRoute . '?' . http_build_query([
            'token' => $this->token,
            'email' => $email,
            'invite' => true,
        ]);
    } else {
        // Use as route name - determine base URL based on configuration
        $routeUrl = route($customRoute, [
            'token' => $this->token,
            'email' => $email,
            'invite' => true,
        ], false);

        // Choose base URL: panel URL or app.url (for backward compatibility)
        $usePanelUrl = config('filament-invite.use_panel_url', false);
        $baseUrl = $usePanelUrl ? $panel->getUrl() : config('app.url');

        $url = rtrim($baseUrl, '/') . '/' . ltrim($routeUrl, '/');
    }
}
```

### Integration with Filament Panel Configuration

When using the default routing, the package automatically respects your panel's authentication configuration:

```php
// In your panel configuration
use Filament\Panel;

public function panel(Panel $panel): Panel
{
    return $panel
        ->passwordResetRouteSlug('custom-reset') // This will be used automatically
        ->passwordResetRequestRouteSlug('custom-request');
}
```

To use custom routing you have plenty of options:

### Option 1: Custom Route Name
```php
'routes' => [
    'reset' => 'your.custom.route.name',
],
```

### Option 2: Full URL (External System)
```php
'routes' => [
    'reset' => 'https://external-auth.example.com/reset-password',
],
```

### Option 3: Different Domain/Subdomain
```php
'routes' => [
    'reset' => 'https://auth.yourapp.com/password-reset',
],
```

The package automatically detects whether you're providing a route name or a full URL and handles the URL construction accordingly.

### URL Base Configuration

For custom routes, you can choose between using the panel's URL or the application's URL:

```php
// config/filament-invite.php
return [
    'routes' => [
        'reset' => 'your.custom.route.name',
    ],
    
    // URL Configuration
    'use_panel_url' => false, // Default: false (uses app.url for backward compatibility)
];
```

**Options:**
- `'use_panel_url' => false` (default): Uses `config('app.url')` - maintains backward compatibility
- `'use_panel_url' => true`: Uses `$panel->getUrl()`

andreia commented on config/filament-invite.php:
```diff
@@ -5,6 +5,6 @@
     'routes' => [
-        'reset' => 'password.reset',
+        'reset' => null,
     ],
```
We can add a new config option to handle the base URL selection:

```php
// URL Configuration
'use_panel_url' => false, // Use panel URL instead of app.url for custom routes (default: false for backward compatibility)
```

andreia commented on src/Actions/InviteAction.php:
```diff
@@ -43,8 +45,10 @@ protected function setUp(): void
                 if (method_exists($user, 'sendPasswordSetNotification')) {
                     $user->sendPasswordSetNotification($token);
+                } elseif (method_exists($user, 'notify')) {
+                    $user->notify(new SetPassword($token, Filament::getId()));
```
We can use `Filament::getCurrentPanel()?->getId()` instead of `Filament::getId()` for better panel context:

```suggestion
                    $user->notify(new SetPassword($token, Filament::getCurrentPanel()?->getId()));
```

andreia commented on src/Actions/InviteAction.php:
```diff
@@ -43,8 +45,10 @@ protected function setUp(): void
+                    $user->notify(new SetPassword($token, Filament::getId()));
                 } else {
-                    Notification::send($user, new SetPassword($token));
+                    Notification::send($user, new SetPassword($token, Filament::getId()));
```
```suggestion
                    Notification::send($user, new SetPassword($token, Filament::getCurrentPanel()?->getId()));
```

andreia commented on src/Notifications/SetPassword.php:
```diff
@@ -76,11 +72,18 @@ public function toMail($notifiable)
             );
         }
 
-        return $message->action(__('Set Password'), url(config('app.url') . route(config('filament-invite.routes.reset'), [
-            'token' => $this->token,
-            'email' => $email,
-            'invite' => true,
-        ], false)));
+        if (empty(config('filament-invite.routes.reset'))) {
+            $url = Filament::getPanel('admin')->getResetPasswordUrl($this->token, $notifiable, ['invite' => true]);
+        } else {
+            $domain = rtrim(Filament::getPanel($this->filamentPanelId)->getPath(), '/');
+            $url = url($domain . '/' . route(config('filament-invite.routes.reset'), [
+                'token' => $this->token,
+                'email' => $email,
+                'invite' => true,
+            ], false));
+        }
```
The current implementation can be improved to support multiple URL scenarios and backward compatibility:

```suggestion
        // Get the appropriate panel
        $panel = $this->filamentPanelId
            ? Filament::getPanel($this->filamentPanelId)
            : Filament::getDefaultPanel();

        if (empty(config('filament-invite.routes.reset'))) {
            // Use Filament's built-in routing (respects panel's passwordResetRouteSlug configuration)
            $url = $panel->getResetPasswordUrl($this->token, $notifiable, ['invite' => true]);
        } else {
            // Support both custom routes and full URLs for maximum flexibility
            $customRoute = config('filament-invite.routes.reset');

            // Check if it's a full URL (starts with http/https)
            if (str_starts_with($customRoute, 'http')) {
                // Use the full URL directly
                $url = $customRoute . '?' . http_build_query([
                    'token' => $this->token,
                    'email' => $email,
                    'invite' => true,
                ]);
            } else {
                // Use as route name - determine base URL based on configuration
                $routeUrl = route($customRoute, [
                    'token' => $this->token,
                    'email' => $email,
                    'invite' => true,
                ], false);

                // Choose base URL: panel URL or app.url (for backward compatibility)
                $usePanelUrl = config('filament-invite.use_panel_url', false);
                $baseUrl = $usePanelUrl ? $panel->getUrl() : config('app.url');

                $url = rtrim($baseUrl, '/') . '/' . ltrim($routeUrl, '/');
            }
        }
```

andreia reviewed (changes requested):
@andrii-trush Thank you for this PR! :)

The flexible routing system is a great addition, especially for applications where `app.url` differs from the Filament URL.

We could make some enhancements leveraging Filament's existing authentication configuration, and also providing a backward compatibility with the current implementation, like so:

andrii-trush committed:
Update src/Notifications/SetPassword.php

Co-authored-by: Andréia Bohner <andreiabohner@gmail.com>

andrii-trush committed:
Update src/Actions/InviteAction.php

Co-authored-by: Andréia Bohner <andreiabohner@gmail.com>

andrii-trush committed:
Update src/Actions/InviteAction.php

Co-authored-by: Andréia Bohner <andreiabohner@gmail.com>

andrii-trush committed:
Add use_panel_url configuration option

Added configuration option to use panel URL for custom routes.

andrii-trush committed:
Enhance README with routing details for password resets

Updated the README to clarify routing options for password reset URLs, including support for custom routes and full URLs.

andrii-trush commented:
@andreia I have applied all your recommendations. Could You review again?

andreia reviewed (approved):
Thank you @andrii-trush :)
